### PR TITLE
Fixes #18906: Fixes title extraction for vivo.sx.

### LINKED ISSUE #18906: Vivo does not work anymore
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2019.01.17*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [X] I've **verified** and **I assure** that I'm running youtube-dl **2019.01.17**

### Before submitting an *issue* make sure you have:
- [X] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [X] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [X] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://vivo.sx/a33b477703']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.01.17
[debug] Python version 3.7.2 (CPython) - Linux-4.20.3-arch1-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1, rtmpdump 2.4
[debug] Proxy map: {}
[Vivo] a33b477703: Downloading webpage
WARNING: unable to extract title; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/bin/youtube-dl", line 11, in <module>
    load_entry_point('youtube-dl==2019.1.17', 'console_scripts', 'youtube-dl')()
  File "/usr/lib/python3.7/site-packages/youtube_dl/__init__.py", line 472, in main
    _real_main(argv)
  File "/usr/lib/python3.7/site-packages/youtube_dl/__init__.py", line 462, in _real_main
    retcode = ydl.download(all_urls)
  File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 2002, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 793, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/common.py", line 508, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/shared.py", line 26, in _real_extract
    'full:title', webpage, 'title')).decode('utf-8')
  File "/usr/lib/python3.7/base64.py", line 80, in b64decode
    s = _bytes_from_decode_data(s)
  File "/usr/lib/python3.7/base64.py", line 46, in _bytes_from_decode_data
    "string, not %r" % s.__class__.__name__) from None
TypeError: argument should be a bytes-like object or ASCII string, not 'NoneType'
...
<end of log>
```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://www.youtube.com/watch?v=BaW_jenozKc
- Single video: https://youtu.be/BaW_jenozKc
- Playlist: https://www.youtube.com/playlist?list=PL4lCao7KL_QFVb7Iudeipvc2BCavECqzc

Note that **youtube-dl does not support sites dedicated to [copyright infringement](https://github.com/rg3/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free)**. In order for site support request to be accepted all provided example URLs should not violate any copyrights.

---

### Description of your *issue*, suggested solution and other information

Comes up with ANY video from vivo since yesterday, worked before.

Explanation of your *issue* in arbitrary form goes here. Please make sure the [description is worded well enough to be understood](https://github.com/rg3/youtube-dl#is-the-description-of-the-issue-itself-sufficient). Provide as much context and examples as possible.
If work on your *issue* requires account credentials please provide them or explain how one can obtain them.

### PR DESCRIPTION
**Pull request by dhoepfl**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fixes #18906

Download from vivo.sx fails since title extraction no longer works. Since vivo shares the code with shared.sx, I moved the title extraction into a `_extract_title` method to the `SharedBase` class which I the override for vivo.sx.

### CODE REVIEW

File: youtube_dl/extractor/shared.py [hunk 1]
```diff
@@ -86,6 +94,14 @@ class VivoIE(SharedBaseIE):
         },
     }
 
+    def _extract_title(self, webpage):
+        data_title = self._search_regex(
+            r'data-name\s*=\s*(["\'])(?P<title>(?:.(?!\1))*.)\1', webpage,
+            'title', default=None, group='title')
+        if data_title:
+            return unescapeHTML(re.sub(r"\.[a-z0-9]{3,4}$", "", data_title))
+        return None
```
Comment (dstftw) [inline youtube_dl/extractor/shared.py]: Must not be None. Read coding conventions.

File: youtube_dl/extractor/shared.py [hunk 2]
```diff
@@ -86,6 +94,14 @@ class VivoIE(SharedBaseIE):
         },
     }
 
+    def _extract_title(self, webpage):
+        data_title = self._search_regex(
+            r'data-name\s*=\s*(["\'])(?P<title>(?:.(?!\1))*.)\1', webpage,
```
Comment (dstftw) [inline youtube_dl/extractor/shared.py]: Does not work as expected in all cases.

File: youtube_dl/extractor/shared.py [hunk 3]
```diff
@@ -1,6 +1,11 @@
 from __future__ import unicode_literals
 
-from .common import InfoExtractor
+import re
+
+from .common import (
+    InfoExtractor,
+    unescapeHTML
```
Comment (dstftw) [inline youtube_dl/extractor/shared.py]: utils

File: youtube_dl/extractor/shared.py (hunk 1 above)
Comment (dhoepfl) [reply youtube_dl/extractor/shared.py]: Fixed in new commit.

File: youtube_dl/extractor/shared.py (hunk 2 above)
Comment (dhoepfl) [reply youtube_dl/extractor/shared.py]: Now using utils.extract_attributes.

File: youtube_dl/extractor/shared.py (hunk 3 above)
Comment (dhoepfl) [reply youtube_dl/extractor/shared.py]: Fixes.

File: youtube_dl/extractor/shared.py [hunk 4]
```diff
@@ -86,6 +93,12 @@ class VivoIE(SharedBaseIE):
         },
     }
 
+    def _extract_title(self, webpage):
+        stream_content = extract_attributes(self._search_regex(
```
Comment (dstftw) [inline youtube_dl/extractor/shared.py]: There is totally no point in `extract_attributes`. Just fix the regex.

Comment (dhoepfl) [reply youtube_dl/extractor/shared.py]: I did not see what was wrong with the regex, so I looked at how other extractors implement similar problems. To me, `extract_attributes` seemed to be a common pattern.

Comment (dhoepfl) [reply youtube_dl/extractor/shared.py]: @dstftw May I ask you to review my new change? I think I fixed the regex.

**Comment by dhoepfl** [comment]
Hi @dstftw,

is there anything left that I have to change? I think I fixed everything you mentioned. Any chance this fix gets merged?

THX,
   Daniel

**Pull request closed without merging**